com.unity.ml-agents/Runtime/Sensor/RayPerceptionSensor.cs

```diff
@@ -134,103 +285,78 @@ public virtual SensorCompressionType GetCompressionType()
             // along with object distance.
-            int bufferOffset = 0;
-            for (var rayIndex = 0; rayIndex < rayAngles.Count; rayIndex++)
+            var unscaledRayLength = input.rayLength;
+            var unscaledCastRadius = input.castRadius;
+            for (var rayIndex = 0; rayIndex < input.angles.Count; rayIndex++)
             {
-                var angle = rayAngles[rayIndex];
-                Vector3 startPositionLocal, endPositionLocal;
-                if (castType == CastType.Cast3D)
-                {
-                    startPositionLocal = new Vector3(0, startOffset, 0);
-                    endPositionLocal = PolarToCartesian3D(unscaledRayLength, angle);
-                    endPositionLocal.y += endOffset;
-                }
-                else
-                {
-                    // Vector2s here get converted to Vector3s (and back to Vector2s for casting)
-                    startPositionLocal = new Vector2();
-                    endPositionLocal = PolarToCartesian2D(unscaledRayLength, angle);
-                }
-
-                var startPositionWorld = transform.TransformPoint(startPositionLocal);
-                var endPositionWorld = transform.TransformPoint(endPositionLocal);
+                var extents = input.RayExtents(rayIndex);
+                var startPositionWorld = extents.StartPositionWorld;
+                var endPositionWorld = extents.EndPositionWorld;
 
                 var rayDirection = endPositionWorld - startPositionWorld;
                 // If there is non-unity scale, |rayDirection| will be different from rayLength.
                 // We want to use this transformed ray length for determining cast length, hit fraction etc.
                 // We also it to scale up or down the sphere or circle radii
                 var scaledRayLength = rayDirection.magnitude;
                 // Avoid 0/0 if unscaledRayLength is 0
-                var scaledCastRadius = unscaledRayLength > 0 ? unscaledCastRadius * scaledRayLength / unscaledRayLength : unscaledCastRadius;
+                var scaledCastRadius = unscaledRayLength > 0 ?
+                    unscaledCastRadius * scaledRayLength / unscaledRayLength :
+                    unscaledCastRadius;
 
                 // Do the cast and assign the hit information for each detectable object.
-                //     sublist[0           ] <- did hit detectableObjects[0]
-                //     ...
-                //     sublist[numObjects-1] <- did hit detectableObjects[numObjects-1]
-                //     sublist[numObjects  ] <- 1 if missed else 0
-                //     sublist[numObjects+1] <- hit fraction (or 1 if no hit)
-
                 bool castHit;
                 float hitFraction;
                 GameObject hitObject;
 
-                if (castType == CastType.Cast3D)
+                if (input.castType == RayPerceptionCastType.Cast3D)
```

Comment:
Also debating moving the raycast logic here to a helper function. Any preference?